File: png2cut.go
package main

import (
	"flag"
	"fmt"
	"image"
	"image/color"
	"image/draw"
	"image/png"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

func hasAlpha(img image.Image) bool {
	switch img.(type) {
	case *image.NRGBA, *image.NRGBA64, *image.RGBA, *image.RGBA64:
		return true
	default:
		return false
	}
}

func padPNGsInDir(targetDir string, saveSubdir string) (int, error) {
	info, err := os.Stat(targetDir)
	if err != nil {
		return 0, err
	}
	if !info.IsDir() {
		return 0, fmt.Errorf("not a directory: %s", targetDir)
	}

	ents, err := os.ReadDir(targetDir)
	if err != nil {
		return 0, err
	}

	var files []string
	for _, e := range ents {
		if e.IsDir() {
			continue
		}
		name := e.Name()
		if strings.HasSuffix(strings.ToLower(name), ".png") {
			files = append(files, name)
		}
	}
	if len(files) == 0 {
		return 0, nil
	}
	sort.Strings(files)

	outDir := filepath.Join(targetDir, saveSubdir)
	if err := os.MkdirAll(outDir, 0o755); err != nil {
		return 0, err
	}

	saved := 0
	for _, fname := range files {
		srcPath := filepath.Join(targetDir, fname)
		f, err := os.Open(srcPath)
		if err != nil {
			continue
		}
		img, err := png.Decode(f)
		f.Close()
		if err != nil {
			continue
		}

		b := img.Bounds()
		w := b.Dx()
		h := b.Dy()
		newW := w + (w % 2)
		newH := h + (h % 2)

		outPath := filepath.Join(outDir, fname)
		if newW == w && newH == h {
			// 直接复制文件以保持原样（避免重新编码带来的细微变化）
			// 如果复制失败，再尝试用 png.Encode 写入
			srcBytes, err := os.ReadFile(srcPath)
			if err == nil {
				_ = os.WriteFile(outPath, srcBytes, 0o644)
				saved++
				continue
			}
			// 回退：重新编码并保存
			of, err := os.Create(outPath)
			if err != nil {
				continue
			}
			_ = png.Encode(of, img)
			of.Close()
			saved++
			continue
		}

		var dst draw.Image
		if hasAlpha(img) {
			dst = image.NewNRGBA(image.Rect(0, 0, newW, newH))
			// 默认零值就是透明，直接复制到左上角
			draw.Draw(dst, image.Rect(0, 0, w, h), img, b.Min, draw.Src)
		} else {
			// 无透明通道：用不透明黑色填充背景，再把原图覆盖上去
			dstRGBA := image.NewRGBA(image.Rect(0, 0, newW, newH))
			black := &image.Uniform{C: color.RGBA{0, 0, 0, 255}}
			draw.Draw(dstRGBA, dstRGBA.Bounds(), black, image.Point{}, draw.Src)
			draw.Draw(dstRGBA, image.Rect(0, 0, w, h), img, b.Min, draw.Src)
			dst = dstRGBA
		}

		of, err := os.Create(outPath)
		if err != nil {
			continue
		}
		_ = png.Encode(of, dst)
		of.Close()
		saved++
	}

	return saved, nil
}

func main() {
	dir := flag.String("dir", ".", "目标目录 (默认 当前目录)")
	out := flag.String("out", "Texture", "输出子目录名")
	flag.Parse()

	n, err := padPNGsInDir(*dir, *out)
	if err != nil {
		fmt.Fprintln(os.Stderr, "错误:", err)
		os.Exit(1)
	}
	fmt.Printf("处理完成：%d 个 PNG 文件已保存到子目录 %q\n", n, *out)
}

File: png2cut.py

import os
from PIL import Image
import argparse


def pad_pngs_in_dir(target_dir: str, save_subdir_name: str = "Texture") -> int:
    """遍历 target_dir（非递归），把目录下的每个 PNG 补齐到偶数宽高并保存到同一目录下的 `save_subdir_name` 子目录。

    返回处理的文件数量。
    """
    target_dir = os.path.abspath(target_dir)
    if not os.path.isdir(target_dir):
        raise NotADirectoryError(target_dir)

    saved = 0
    # 列出目录中的 png 文件并排序，避免不稳定顺序
    files = sorted([f for f in os.listdir(target_dir) if f.lower().endswith('.png')])
    if not files:
        return 0

    out_dir = os.path.join(target_dir, save_subdir_name)
    os.makedirs(out_dir, exist_ok=True)

    for fname in files:
        src_path = os.path.join(target_dir, fname)
        try:
            img = Image.open(src_path)
        except Exception:
            # 跳过无法打开的文件
            continue

        w, h = img.size
        new_w = w + (w % 2)
        new_h = h + (h % 2)

        if new_w == w and new_h == h:
            # 已经是偶数，不需要修改，直接复制到输出目录
            img.save(os.path.join(out_dir, fname))
            saved += 1
            continue

        # 选择填充色：RGBA 用透明 (0,0,0,0)，其他模式用 0
        mode = img.mode
        if 'A' in mode:
            fill = (0,) * len(img.getbands())
        else:
            fill = 0

        new_img = Image.new(mode, (new_w, new_h), fill)
        new_img.paste(img, (0, 0))
        new_img.save(os.path.join(out_dir, fname))
        saved += 1

    return saved


def main():
    parser = argparse.ArgumentParser(description='Pad PNG files in a directory to even width/height and save to a Texture subfolder.')
    parser.add_argument('dir', nargs='?', default='.', help='目标目录 (默认: 当前目录)')
    args = parser.parse_args()

    count = pad_pngs_in_dir(args.dir)
    print(f"处理完成：{count} 个 PNG 文件已保存到子目录 'Texture'。")


if __name__ == '__main__':
    main()